internal/core/src/exec/operator/query-agg/GroupingSet.cpp

```diff
@@ -0,0 +1,248 @@
+    int32_t accumulatorFlagsOffset = 0;
+    int32_t alignment = 1;
+
+    for (auto& aggregate : aggregates_) {
+        auto& function = aggregate.function_;
+        Accumulator accumulator(function.get());
+        // Accumulator offset must be aligned by their alignment size.
+        offset = milvus::bits::roundUp(offset, accumulator.alignment());
+        function->setOffsets(offset,
+                             RowContainer::nullByte(accumulatorFlagsOffset),
+                             RowContainer::nullMask(accumulatorFlagsOffset),
+                             rowSizeOffset);
+        offset += accumulator.fixedWidthSize();
+        accumulatorFlagsOffset += 1;
+        alignment =
+            RowContainer::combineAlignments(accumulator.alignment(), alignment);
+    }
+    AssertInfo(__builtin_popcount(alignment) == 1,
+               "alignment of aggregations must be power of two");
+    offset = milvus::Align(offset, alignment);
+    // Use unique_ptr for exception-safe memory management
+    // If initializeNewGroups throws, the buffer will be automatically cleaned up
+    globalAggregationBuffer_ = std::make_unique<char[]>(offset);
+    lookup_->hits_[0] = globalAggregationBuffer_.get();
+    const auto singleGroup = std::vector<vector_size_t>{0};
+    for (auto& aggregate : aggregates_) {
+        aggregate.function_->initializeNewGroups(lookup_->hits_.data(),
+                                                 singleGroup);
+    }
+    globalAggregationInitialized_ = true;
+}
+
+void
+GroupingSet::addGlobalAggregationInput(const milvus::RowVectorPtr& input) {
+    initializeGlobalAggregation();
+    auto* group = lookup_->hits_[0];
+    for (auto i = 0; i < aggregates_.size(); i++) {
+        auto& function = aggregates_[i].function_;
+        populateTempVectors(i, input);
+        function->addSingleGroupRawInput(group, tempVectors_);
+    }
+    tempVectors_.clear();
+}
+
+bool
+GroupingSet::getGlobalAggregationOutput(milvus::RowVectorPtr& result) {
+    initializeGlobalAggregation();  // when input from upstream operator is empty, we need to initialize the accumulators for global aggregation
+    AssertInfo(lookup_->hits_.size() == 1,
+               "GlobalAggregation should have exactly one output line");
+    auto groups = lookup_->hits_.data();
+    for (auto i = 0; i < aggregates_.size(); i++) {
+        auto& function = aggregates_[i].function_;
+        auto resultVector = result->child(aggregates_[i].output_);
+        function->extractValues(groups, 1, &resultVector);
+    }
+    return true;
+}
+
+bool
+GroupingSet::getOutput(milvus::RowVectorPtr& result) {
+    if (isGlobal_) {
+        return getGlobalAggregationOutput(result);
+    }
+    AssertInfo(hash_table_ != nullptr,
+               "hash_table_ should not be nullptr for non-global aggregation");
+    const auto& all_rows = hash_table_->rows()->allRows();
+    if (!all_rows.empty()) {
+        extractGroups(result);
```

Comment:
can we extractGroups in streaming mode to avoid accumulate too many data?

We can leave this to the next pr